Builder pattern for convenient construction of SQLPipeline[Statements]

mrzzzrm opened the pull request:
```c++
/**
 * Interface for the configured execution of SQL.
 *
 * Minimal usage:
 *      SQL{"SELECT * FROM t;"}.pipeline().get_result_table()
 *
 * With custom Optimizer and TransactionContext:
 *      SQL{query}.
 *          set_optimizer(optimizer).
 *          set_transaction_context(tc).
 *          pipeline();
 *
 * Defaults:
 *  - MVCC is enabled
 *  - The default Optimizer (Optimizer::create_default_optimizer() is used.
 *
 * Favour this interface over calling the SQLPipeline[Statement] constructors with their long parameter list.
 * See SQLPipeline[Statement] doc for these classes, in short SQLPipeline ist for queries with multiple statement,
 * SQLPipelineStatement for single statement queries.
 */
```

This stems from my MA where I need to introduce more options to the SQLPipeline and the current way of adding constructors just doesn't scale. The name `SQL` may seem somewhat provocatively, but I think it works nicely as an interface. If you have better ideas for names, let me know.

mrks commented:
Have you considered a settings-struct with defaults and designated initializers?

mrzzzrm commented:
I think `class SQL` is essentially that, just a bit more convenient to call (and we do that in a lot of places). Settings struct would look like this:

```c++
SQLPipelineDesc desc{sql_string};
desc.use_mvcc = UseMvcc::No; // or setter, idc
SQLPipeline{desc}.//do_things();
```
or 
```c++
SQLPipeline{SQLPipelineDesc desc{sql_string}.set_use_mvcc(UseMvcc::No)}.//do_things();
```
vs the proposed
```c++
SQL{sql_string}.set_use_mvcc(UseMvcc::No}.//do_things();
```

I think this proposal is pretty convenient to use. I want it to be like this because its used often (mostly in Tests, console, tpch_test, the SQL testrunner, tpch benchmark) AND because its something you can show to newcomers like "look how easy it is to execute SQL in hyrise".

mrks commented:
```
SQLPipeline{“SELECT ...”, {.optimizer = super_optimizer,
                           .transaction_context = tx}}.get_result_table()
```

mrzzzrm commented:
I think that's a C thing that doesn't work in C++?

mrzzzrm commented:
![image](https://user-images.githubusercontent.com/2679018/37343076-b664e17c-26c6-11e8-852b-619b939f61f6.png)

mjendruk commented on src/lib/sql/sql.hpp:
```diff
@@ -0,0 +1,61 @@
+ private:
+  const std::string _sql;
+
+  UseMvcc _use_mvcc{UseMvcc::Yes};
```
Are we generally okay with defining default values here?

mrzzzrm replied:
I am and I think we do it in places.

mjendruk commented on src/lib/sql/sql_pipeline.hpp:
```diff
@@ -32,30 +32,9 @@ namespace opossum {
-
-  SQLPipeline(const std::string& sql, const std::shared_ptr<Optimizer>& optimizer,
-              const PreparedStatementCache& prepared_statements,
-              std::shared_ptr<TransactionContext> transaction_context);
```
When I look at all these constructors, it seems absolutely clear, why you wanted something else.

mjendruk commented on src/lib/sql/sql_pipeline.hpp:
```diff
@@ -116,13 +92,14 @@ class SQLPipeline : public Noncopyable {
   // e.g. CREATE VIEW foo AS SELECT * FROM bar; SELECT * FROM foo;
   // --> requires execution of first statement before the second one can be translated
-  bool _requires_execution;
+  bool _requires_execution{false};
```
One question: Here you use the braces but in constructors, you use parantheses. Is there a reason for this?

mrzzzrm replied:
because I'm bound by my own habits and I don't tend to question them ;)

mrks commented:
So? According to [this](https://gcc.gnu.org/onlinedocs/gcc/C-Dialect-Options.html) and [this](https://clang.llvm.org/cxx_status.html) we should be able to use `-std=c++2a`.

mjendruk commented on src/lib/sql/sql.hpp:
```diff
@@ -0,0 +1,61 @@
+  SQL& disable_mvcc();
+
+  SQLPipeline pipeline() const;
+  SQLPipelineStatement pipeline_statement() const;
```
So an SQLPipelineStatements represents a single SQL statement and a pipeline can consist of several, am I right?

mrzzzrm replied:
yep

mjendruk replied:
So what happens if you call `pipeline_statement` if the passed SQL string contains several statements?

mrzzzrm replied:
What happened before. SQLPipelineStatement will go down with a poof. Gracefully. That means it asserts.

mrzzzrm replied:
```
  Assert(_parsed_sql_statement->size() == 1,
         "SQLPipelineStatement must hold exactly one statement. "
         "Use SQLPipeline when you have multiple statements.");
```

mrzzzrm commented:
Mmh, while I find aggregate initializers to be a valuable addition to C++ I don't think upping our language level (to experimental!) for this is wise. Might run into platform specific problems and people might start using other c++20 features as well. I'm not even convinced that it makes writing a SQL call with custom options more convenient (curly braces, unclear layouting as in your example, spaces, dots, etc.).

mjendruk commented:
Some more thoughts: With a simple struct, it would be still possible to initialise it without names, so we’d have to write somewhere, that it should be initialised using this new style.

mrzzzrm commented:
@mrks looking at your two links again, I can't see support for aggregate initializers (www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0329r0.pdf) in "C++2a implementation status".

lawben commented on src/lib/sql/sql.cpp:
```diff
@@ -0,0 +1,43 @@
+SQLPipelineStatement SQL::pipeline_statement() const {
+  auto optimizer = _optimizer ? _optimizer : Optimizer::create_default_optimizer();
+
+  return {_sql, nullptr, _use_mvcc, _transaction_context, optimizer, _prepared_statements};
```
could you comment the `nullptr` here. i guess this is the parse result that we don't have..?

mrzzzrm commented:
@mjendruk you disapprove because I use {}-initialization here and =-initialization there?

mrks commented:
Ok, let’s leave designated initializers out for now. I still think that if both compilers support it, we should switch to `c++2a` soon. `c++1z` didn’t give us any problems (the gcc bug that we found was `c++14`, I believe).

mrks commented:
About the PR: Could you explain why we need a separate class for this instead of just setting the options in the pipeline? I like the pipeline as it is and feel like the name is appropriate for the entity that we put SQL in and get tables out. We could create a pipeline with a SQL string, set all options, and then call `get_result_table`. If the pipeline was already started and someone tries to set an option, we could throw an exception.

mrzzzrm commented:
It's the old question of static correctness vs verifying correctness at runtime. I much prefer the former.

mrzzzrm commented:
Also, if we'd follow your proposal we'd duplicate these checks in both SQLPipeline and SQLPipelineStatement.

mrks commented:
`set_...` could check if the pipeline has already been used, so there is no need to touch `SQLPipelineStatement`.

I feel like having a `SQLPipelineBuilder` and a `SQLPipeline` contradicts your idea of showing newcomers how easy it is to execute an SQL statement.

mrzzzrm commented:
You **can** use the SQLPipelineStatement directly, so you'd need to do the check there as well. I share your sentiments towards having an additional class adding complexity, but I think it's better than rules about which functions on an object can be called in which order that need to be checked at runtime.

mjendruk commented:
I like the separation of concerns, especially if as Moritz said, there are going to be more options in the future.

mrks commented:
Ok, we can do it this way. I don‘t like the name, though. First, it doesn‘t actually tell anyone what it does. Second, it is not the core component for SQL evaluation. Third, calling `set_use_mvcc` on an `SQL` class is wrong - SQL has no notion of MVCC (or optimizers, or ...).

I would consider `SQLPipelineBuilder`.

mrks commented on src/lib/sql/sql.cpp:
```diff
@@ -0,0 +1,43 @@
+  return *this;
+}
+
+SQL& SQL::disable_mvcc() { return set_use_mvcc(UseMvcc::No); }
```
Why do we have this and `set_use_mccc`?

mrzzzrm replied:
disable_mvcc() is done a lot, so it seemed convenient. Will remove it if you want to. Will doc it if you want to.

mrks commented on src/bin/console/console.cpp:
```diff
@@ -220,9 +221,14 @@ int Console::_eval_command(const CommandFunction& func, const std::string& comma
     if (_explicitly_created_transaction_context != nullptr) {
-      _sql_pipeline = std::make_unique<SQLPipeline>(sql, _prepared_statements, _explicitly_created_transaction_context);
+      _sql_pipeline =
+          std::make_unique<SQLPipeline>(SQL{sql}
```
Aren‘t you copying the pipeline here?

mrzzzrm replied:
Moving it, if I'm not mistaken

mrks commented on src/lib/sql/sql.hpp:
```diff
@@ -0,0 +1,66 @@
+ * Interface for the configured execution of SQL.
+ *
+ * Minimal usage:
+ *      SQL{"SELECT * FROM t;"}.pipeline().get_result_table()
```
To be honest, I am not a fan of that chained syntax. :/
It is mostly uncommon in C++ and it is unclear on what you are operating `SQL{...}.set_optimizer(...).pipeline().get_result_table()` has three objects that you operate on. We don‘t save anything by chaining it, except for maybe two line breaks.

mrzzzrm replied:
I have seen these chained setters to be prefixed with `with_`, so it would be `SQL{"..."}.with_mvcc_disabled().with_transaction_context(context).create_pipeline()`

mrzzzrm replied:
We save typing `sql_pipeline_builder.` over and over again and its respective linebreaks. I think not chaining this would feel reeeeaaally awkward. (and Java-esque)

mrks replied:
„code is read far more often than it is written“
https://softwareengineering.stackexchange.com/questions/80244/are-there-any-actual-drawbacks-to-self-referential-method-chaining

I find this harder to follow than one-thing-per-line. It gets worse when things span over multiple lines and the linter dows weird things.

lawben replied:
I'm also for option 1 :) I think this is a very common approach when using a builder.

mrks commented on src/lib/sql/sql.hpp:
```diff
@@ -0,0 +1,66 @@
+ *  - MVCC is enabled
+ *  - The default Optimizer (Optimizer::create_default_optimizer() is used.
+ *
+ * Favour this interface over calling the SQLPipeline[Statement] constructors with their long parameter list.
```
Do we want this to be the one-and-only way to execute SQL? Maybe the `SQLPipeline` constructor should be protected and this class friended?

mrzzzrm replied:
I'm not the biggest fan of forbidding object creation to enforce preferred usage patterns. Will not fight over it in this case though. Your call.

mrks replied:
Nope, just an idea

mrks commented on src/lib/sql/sql_pipeline.hpp:
```diff
@@ -100,9 +79,6 @@ class SQLPipeline : public Noncopyable {
-
   std::vector<std::shared_ptr<SQLPipelineStatement>> _sql_pipeline_statements;
 
   std::shared_ptr<TransactionContext> _transaction_context;
```
Now we should be able to const most of these, shouldn‘t we?

mrzzzrm commented:
Naming of `SQL`. I don't want the term "Builder" in the name. It's just daunting to read and a lot to type. I agree that `SQL` is not the best name, but if we switch to the `with_` setters it's actually logical to read. "Create a pipeline for this SQL query that includes MVCC and uses the following optimizer...".

What do you think of `Query`, `QueryPipeline` and `QueryPipelineStatement`?

mrzzzrm commented:
That being said, I will go with `SQLPipelineBuilder` if necessary, if we can't find a better name.

mrks commented:
If it behaves like a builder and quacks like a builder, [it probably is one](https://www.javaworld.com/article/2074938/core-java/too-many-parameters-in-java-methods-part-3-builder-pattern.html).

mrzzzrm commented:
* Changed name to `SQLPipelineBuilder`
 * Changed setters to `with_*()`-style
 * Changed creators to `create_pipeline()`

also, I had another branch with the same prefix as this one lying around and did the changes there at first. Hopefully the merge didn't screw things up.

mrks commented:
Thanks!
